external/js/cothority/src/byzcoin/contracts/credentials-instance.ts

```diff
@@ -124,16 +123,20 @@ export class CredentialStruct extends Message<CredentialStruct> {
         registerMessage("personhood.CredentialStruct", CredentialStruct, Credential);
     }
 
-    public credentials: Credential[];
+    /**
+     * Returns a new CredentialStruct class
+     * @param d
+     */
+    static fromData(d: Buffer): CredentialStruct {
+        return CredentialStruct.decode(d);
+    }
+
+    readonly credentials: Credential[];
```

Comment:
I don't understand how that works - how can `credentials` be readonly, and then we do this?

https://github.com/dedis/cothority/blob/831f24277d778cafc78611ba27985c4e8f00d0d5/external/js/cothority/src/byzcoin/contracts/credentials-instance.ts#L86

Reply:
_readonly_ only prevents you from doing `this.credentials = []` but it doesn't prevent you from updating the array. I want to avoid people from removing an entire array accidentally and then force them to use specific API like push/splice so you know that you have previous data to deal with. If we want something more strict we can freeze the array so push will be forbidden and you would need to instantiate a new struct with the new credentials.